src/main/kotlin/no/skatteetaten/aurora/boober/service/resourceprovisioning/S3StorageGridProvisioner.kt

```diff
@@ -0,0 +1,189 @@
+import mu.KotlinLogging
+import no.skatteetaten.aurora.boober.feature.OperationScopeFeature
+import no.skatteetaten.aurora.boober.model.openshift.*
+import no.skatteetaten.aurora.boober.service.*
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftClient
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftDeployer
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftResponse
+import no.skatteetaten.aurora.boober.service.openshift.OperationType
+import no.skatteetaten.aurora.boober.utils.convert
+import no.skatteetaten.aurora.boober.utils.normalizeLabels
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Service
+
+data class SgProvisioningRequest(
+    val tenant: String,
+    val objectAreaName: String,
+    val deploymentName: String,
+    val targetNamespace: String,
+    val bucketPostfix: String
+)
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+data class StorageGridCredentials(
+    val tenantName: String,
+    val serviceEndpoint: String,
+    val accessKey: String,
+    val secretKey: String,
+    val bucketName: String,
+    val objectPrefix: String,
+    val username: String,
+    val password: String,
+    val bucketRegion: String? = null
+)
+
+data class SgRequestsWithCredentials(
+    val request: SgProvisioningRequest,
+    val credentials: StorageGridCredentials
+)
+
+class S3ProvisioningException(message: String, cause: Throwable? = null) : ProvisioningException(message, cause)
+
+@Service
+@ConditionalOnProperty(value = ["integrations.s3.variant"], havingValue = "storagegrid", matchIfMissing = false)
+class S3StorageGridProvisioner(
+    val openShiftDeployer: OpenShiftDeployer,
+    val openShiftClient: OpenShiftClient,
+    val openShiftCommandService: OpenShiftCommandService,
+    val herkimerService: HerkimerService,
+    val operationScopeFeature: OperationScopeFeature,
+    @Value("\${integrations.storagegrid.bucket.region:us-east-1}") val defaultBucketRegion: String,
+    @Value("\${integrations.storagegrid.provisioning-timeout:20000}") val provisioningTimeout: Long,
+    @Value("\${integrations.storagegrid.provisioning-status-check-interval:1000}") val statusCheckIntervalMillis: Long
+) {
+    private val logger = KotlinLogging.logger { }
+
+    fun getOrProvisionCredentials(applicationDeploymentId: String, requests: List<SgProvisioningRequest>)
+            : List<SgRequestsWithCredentials> {
+
+        if (requests.isEmpty()) return emptyList()
+
+        provisionMissingObjectAreas(applicationDeploymentId, requests)
+        return herkimerService.getObjectAreaAdminCredentials(applicationDeploymentId, requests)
+    }
+
+    fun provisionMissingObjectAreas(applicationDeploymentId: String, requests: List<SgProvisioningRequest>) {
+        val existingResources = herkimerService.getObjectAreaResourcesIndex(applicationDeploymentId)
```

Comment:
Jeg har vel nevnt dette før, men jeg synes at Boober ikke skal sjekke om det eksisterer alt. Er det noen problemer som oppstår med OCP3 hvis man alltid skriver.

Hvis denne skal være her synes jeg den bør løftes ut et hakk. Metode navnet er missvisende i forhold til hva som gjøres her. Den er mer korrekt et nivå opp. Altså getOrProvisionCredentials

Reply:
Tja, tjo... dersom provisjonering har lykkes (credentials eksisterer) er det jo ikke stort poeng - egentlig - å prøve provisjonere på nytt, og evt. vente. Men vi kan gjøre det altså. Men greit om vi kan ta det etter en merge.

Men jeg skjønner ikke helt hva du mener med at navnet er missvisende? For å provisjonere de som mangler må man først hente ut de som allerede eksisterer, fjerne requestene for de som eksisterer, og deretter provisjonere de som mangler. Denne metoden, "getter" jo egentlig ingenting, sånn sett. Ikke for andre enn seg selv. "Gettingen" av credentials gjøres når provisjoneringen er ferdig - altså at denne metoden er ferdig kjørt.

Reply:
Er det alltid slik at hvis det finnes i herkimer så finnes det en tilsvarende SGOA ressurs eller er det unntak? Det er det eneste jeg kunne tenkt meg fordelen var med alltid å skrive.

Jeg ser nå at jeg tok feil angående at det var missvissende navn. Du har helt korrekt. Jeg mistolket det med at du sjekket etter missing resources som at du getted, men du returnerer de jo aldri ut av metoden.